src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,433 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package javax.crypto.spec;
+
+import jdk.internal.javac.PreviewFeature;
+
+import javax.crypto.SecretKey;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Parameters for the combined Extract-Only, Expand-Only, or Extract-then-Expand
+ * operations of the HMAC-based Key Derivation Function (HKDF). The HKDF
+ * function is defined in <a href="http://tools.ietf.org/html/rfc5869">RFC
+ * 5869</a>.
+ *
+ * @since 23
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public interface HKDFParameterSpec extends KDFParameterSpec {
+
+    /**
+     * This builder helps with the mutation required by the {@code Extract}
+     * scenario.
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+    final class Builder {
+
+        Extract extract = null;
+        List<SecretKey> ikms = new ArrayList<>();
+        List<SecretKey> salts = new ArrayList<>();
+
+        Builder() {}
+
+        /**
+         * Creates a {@code Builder} for an {@code Extract}.
+         *
+         * @return a {@code Builder} to mutate
+         */
+        private Builder createExtract() {
+            extract = new Extract();
+            return this;
+        }
+
+        /**
+         * Akin to a {@code Builder.build()} method for an {@code Extract}
+         *
+         * @return an immutable {@code Extract}
+         */
+        public Extract extractOnly() {
+            if (this.ikms.isEmpty() && this.salts.isEmpty()) {
+                throw new IllegalStateException(
+                    "this `Builder` must have either at least one IKM value, "
+                    + "at least one salt "
+                    + "value, or values for both before calling `extractOnly`");
+            } else {
+                this.extract = new Extract(ikms, salts);
+                return this.extract;
+            }
+        }
+
+        /**
+         * Akin to a {@code Builder.build()} method for an
+         * {@code ExtractExpand}
+         *
+         * @param info
+         *     the optional context and application specific information
+         * @param length
+         *     the length of the output key material
+         *
+         * @return an {@code ExtractExpand}
+         */
+        public ExtractExpand andExpand(byte[] info, int length) {
+            if (extract == null) {
+                throw new IllegalStateException(
+                    "`andExpand` can only be called on a `Builder` when "
+                    + "`ofTypeExtract` has "
+                    + "been called");
+            }
+            return new ExtractExpand(
+                extractOnly(), info,
+                length);
+        }
+
+        /**
```

Comment:
The method spec should start with a paragraph saying "Adds key input material (IKM) to this builder", and then a new paragraph can talk about "AddIKM may be called....".

Also, the "may be called when..." sounds like there is another `setIKM` method when the IKM can be provided in a single shot. We should be very clear that this is the only method. We can probably say something like this:

This method can be invoked multiple times on a builder, appending each newly added IKM to the existing IKM. This is particularly useful for "labeled" HKDF Extract used in TLS 1.3 and HPKE, where the IKM consists of concatenated components, which may include both byte arrays and (possibly non-extractable) secret keys.

This applies to the 3 methods below as well.